### DISCUSSION TITLE
How to include utility methods in actor.attemptsTo()

### DISCUSSION THREAD

**Discussion by Krithikmano** [serenity-js/serenity-js#1638] [How to... ?]
I have a utility class which has multiple methods and most of methods return types are string . I want to call those method in actor.actor.attemptsTo(Enter.theValue(here I have call utility method ).into.....

When I try to call it is showing as  required type ‘Activity’
Please help out how to do that and provide some example.

**Comment by jan-molak** [comment]
Hey @Krithikmano! It would be easier to discuss this over code. Could you please provide a more concrete example of such a utility method and how you'd like to use it? Thanks!

**Comment by Krithikmano** [comment]
```typescript
export class FligtData {

    public async getRoute(fligtName: string): Promise<string>
        fName:any=null
        const fligtInstance = axios.create({
            baseURL: https://example.com,
        });
        const actor = actorCalled("flight").whoCan(TakeNotes.usingAnEmptyNotepad(),CallAnApi.using(fligtInstance));
        await actor.attemptsTo(
            Send.a(GetRequest.to(/fligtName)),
            TakeNote.of(LastResponse.body()),
            Log.the(LastResponse.body()),
            Ensure.that(LastResponse.status(), equals(200))
        );
        fName = (await Note.of(LastResponse.body()).answeredBy(actor)); //storing response in fName variable
        return fName.city; // here extract the city from fName response
    }

}
```

I want to call the above method in the below class
```typescript
const fData = new FligtData()
export class flightPage{
public async validateFlightPage() {
    actor: any = actorCalled('TravelMemeber').whoCan(TakeNotes.usingAnEmptyNotepad(), BrowseTheWeb.using(protractor.browser))
    for (var name of fligtName) { //fligtName is array
        await this.actor.attemptsTo(Task.where('validate TravelPage',
        Enter.theValue(fData.getRoute(name)).into(TargetElemntFlightName), //here I am getting error while entering this method fData.getRoute(name)
        .....
        Click.on(TargetElementsubmit),
    }
  }
}
```

**Comment by jan-molak** [comment]
Thanks, it makes it a bit more clear. It feels like there's an easier way to accomplish what you're after and perhaps the issue could be a result of mixing different implementation styles.

To begin with, you'd want to create a class that instantiates your actors to avoid doing it on every call, for example:

```typescript
import { Actor, Cast, TakeNotes } from '@serenity-js/core';
import { CallAnApi } from '@serenity-js/rest';

export class Actors implements Cast {
    prepare(actor: Actor): Actor {
        switch (actor.name) {
            case 'flight': 
                return actor.whoCan(
                    TakeNotes.usingAnEmptyNotepad(),
                    CallAnApi.using(axios.create({ baseUrl: '...' }))
                );
            case 'TravelMember':
                return actor.whoCan(/* etc */)
            default:
                return actor;
        }
    }
}
```

Have a look at how I do it in one of the project templates, like [this one](https://github.com/serenity-js/serenity-js-mocha-protractor-template).

Next, you can replace this logic:
```typescript
fName:any=null
// ...
fName = (await Note.of(LastResponse.body()).answeredBy(actor)); //storing response in fName variable
return fName.city; // here extract the city from fName response
```
with a mapping function:
```typescript
Note.of(LastResponse.body()).map(actor => body => body.city)
```
Not to mention that you can make the actor take note of just the `city` param, if that's the only thing you are interested in:

```typescript
TakeNote.of(LastResponse.body()).map(actor => body => body.city)).as('city'),
```

which you can then retrieve as follows:
```typescript
Note.of('city')
```

The `for` loop could be replaced with a task to [`Loop`](https://serenity-js.org/modules/core/class/src/screenplay/tasks/Loop.ts~Loop.html):

```typescript
actorCalled('TravelMemeber').attemptsTo(
    Loop.over(flightNames).to(  // I'm guessing it's a list of flight names in your code sample?
         Enter.theValue(Loop.item<string>()).into(/*...*/),
         // etc
     )
)
```

But, given that the only thing `FligtData.getRoute` seems to be doing is to make a `GET` call and fetch some data, you could replace it with a custom question, i.e.:

```typescript
const FlightRoute = (flightName: Question<string>) => 
    Question.about(`flight route for ${ flightName }`, actor => {
        return CallAnApi.as(actor).request(/* AxiosRequest */).then(data => data.city)
    })
```

Does this help?

**Comment by Krithikmano** [comment]
Thanks for the information. But, I am bit confused to implement with your style. I am beginner. However, I am following my style and I could call utility method successfully as below approach (included Quesiton.about and assign the getRoute method to route variable). See the code below.

In the first iteration it is working fine, the second iteration I am getting the error like "ConfigurationError: flight already has an ability to TakeNotesUsingAnEmptyNotepad, so you don't need to give it to them again.". How to resolve this error

const fData = new FligtData()
export class flightPage{

route: any = (flightName:string) => Question.about('To get city name', actor => fData.getRoute(flightName))

public async validateFlightPage() {
    actor: any = actorCalled('TravelMemeber').whoCan(TakeNotes.usingAnEmptyNotepad(), BrowseTheWeb.using(protractor.browser))
    for (var name of fligtName) { //fligtName is array
        await this.actor.attemptsTo(Task.where('validate TravelPage',
        Enter.theValue(this.route(name)).into(TargetElementName), //here I am getting error while entering this method fData.getRoute(name)
        .....
        Click.on(TargetElementsubmit),
    }
  }
}

**Comment by jan-molak** [comment]
Hey @Krithikmano,

No worries, in the below error:
```
ConfigurationError: flight already has an ability to TakeNotesUsingAnEmptyNotepad, so you don't need to give it to them again
```
Serenity/JS tells you that there's a configuration problem with an actor called "`flight`", and that it seems like you're trying to configure this actor more than once with the same ability called `TakeNotesUsingAnEmptyNotepad`.

An actor requires to have **at most one** instance of a given ability, and using `actorCalled(name)` makes Serenity/JS "remember" the actor, together with any abilities you've given it.

This means that the below code sample would result in an error you're seeing:
```typescript
actorCalled('Bob').whoCan(TakeNotes.usingAnEmptyNotepad())
actorCalled('Bob').whoCan(TakeNotes.usingAnEmptyNotepad()) // ConfigurationError
```

The way to avoid this is to **separate the concerns** of configuring the actors from performing the test.

To do this, you can create a class that implements an interface called [`Cast`](https://serenity-js.org/modules/core/class/src/stage/Cast.ts~Cast.html), tell Serenity/JS to use it (see [`engage`](https://serenity-js.org/modules/core/function/index.html#static-function-engage)), then in your test code refer to your actors by name (i.e. [`actorCalled('Bob')`](https://serenity-js.org/modules/core/function/index.html#static-function-actorCalled), or [`actorInTheSpotlight()`](https://serenity-js.org/modules/core/function/index.html#static-function-actorInTheSpotlight) to get the most recently accessed actor), and **avoid calling `whoCan`** with the same ability on the same actor more than once.

Does this help?
